### PR TITLE
Kernel/Net: Implement SO_SNDBUF and SO_RCVBUF for AF_LOCAL

### PR DESCRIPTION
**Pull request by logkos**
These options are generally used for QoS tuning by Apache, NGINX, MySQL etc. (insert anything that scales here)

### CODE REVIEW

File: Kernel/Net/LocalSocket.cpp [hunk 1]
```diff
@@ -399,14 +399,27 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
     MutexLocker locker(mutex());
 
+    size_t buffer_size;
     socklen_t size;
     TRY(copy_from_user(&size, value_size.unsafe_userspace_ptr()));
 
     switch (option) {
     case SO_SNDBUF:
-        return ENOTSUP;
+        if (size < sizeof(size_t))
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:408]: This can be !=

File: Kernel/Net/LocalSocket.cpp [hunk 2]
```diff
@@ -399,14 +399,27 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+            return EINVAL;
+        buffer_size = m_for_server.ptr()->space_for_writing();
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:410]: This seems to be the currently available write space, not the entire buffer's size? So not quite correct. Also, you can access all members with a direct arrow operator, e.g. `m_for_server->whatever_method_i_need()`

File: Kernel/Net/LocalSocket.cpp [hunk 3]
```diff
@@ -399,14 +399,27 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+        TRY(copy_to_user(static_ptr_cast<int*>(value), bit_cast<int*>(&buffer_size)));
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:411]: Again, the pointer cast is utterly wrong and the type is size_t. This bit_cast is wrong and will be UB on big-endian systems.

File: Kernel/Net/LocalSocket.cpp [hunk 4]
```diff
@@ -399,14 +399,27 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+        size = sizeof(size_t);
+        TRY(copy_to_user(value_size, &size));
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:413]: You can put the sizeof(size_t) directly in this copy_to_user instruction, no need to save the size variable.

File: Kernel/Net/LocalSocket.cpp [hunk 5]
```diff
@@ -399,14 +399,27 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+        return {};
     case SO_RCVBUF:
-        return ENOTSUP;
+        if (size < sizeof(size_t))
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:416]: (GitHub doesn't let me select multiple lines for review right now)

Same corrections from above apply to SO_RCVBUF as well.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Kernel/Net/LocalSocket.cpp (hunk 3 above)
Comment (logkos) [reply Kernel/Net/LocalSocket.cpp]: what cast should be used here? pointer cast seems to be used on everywhere else

File: Kernel/Net/LocalSocket.cpp [hunk 6]
```diff
@@ -399,14 +399,25 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+    size_t buffer_size;
     socklen_t size;
     TRY(copy_from_user(&size, value_size.unsafe_userspace_ptr()));
 
     switch (option) {
     case SO_SNDBUF:
-        return ENOTSUP;
+        if (size != sizeof(size_t))
+            return EINVAL;
+        buffer_size = m_for_server->immediately_readable();
```
Comment (timschumi) [inline Kernel/Net/LocalSocket.cpp:410]: This appears to return the amount of data that is currently buffered. `SO_SNDBUF` and `SO_RCVBUF` are supposed to return the total buffer size.

**Review by timschumi — APPROVED** [review]

**Pull request merged**